File: backend/src/literature/pdf_text.ts
import { existsSync } from "node:fs";
import { materializeAsset } from "../assets/embedded";
import { resolvePython } from "../simulation/platform";
import PDF_TEXT_PY from "./pdf_text.py" with { type: "text" };

// V66：PDF 全文抽取的 TS 侧。走 V27 验证过的组合修法：脚本内容静态 import 进二进制、
// 运行期解包到真实路径再 spawn（外部 python 读不了 /$bunfs 虚拟路径）。
//
// 失败语义：**永远不抛**——返回 ok:false + 可读 reason。上游（精读卡）据此降级回
// 摘要模式并把降级原因显式写进卡片元数据；抽不出全文不是错误，是要如实标注的事实。

export interface PdfTextResult {
  ok: boolean;
  text?: string;
  pages?: number;
  truncated?: boolean;
  reason?: string;
}

export async function extractPdfText(
  pdfPath: string,
  options: { maxChars?: number; python?: string; timeoutMs?: number } = {},
): Promise<PdfTextResult> {
  const maxChars = options.maxChars ?? 40_000;
  if (!existsSync(pdfPath)) return { ok: false, reason: `PDF 文件不存在: ${pdfPath}` };
  const script = materializeAsset("literature", "pdf_text.py", PDF_TEXT_PY);
  const python = options.python ?? resolvePython();
  try {
    const proc = Bun.spawn([python, script, pdfPath, String(maxChars)], {
      stdout: "pipe",
      stderr: "pipe",
      timeout: options.timeoutMs ?? 60_000,
    });
    const stdout = await new Response(proc.stdout).text();
    const stderr = await new Response(proc.stderr).text();
    await proc.exited;
    const line = stdout.trim().split("\n").pop() ?? "";
    try {
      const parsed = JSON.parse(line) as { ok: boolean; text?: string; pages?: number; truncated?: boolean; error?: string };
      if (parsed.ok && typeof parsed.text === "string") {
        return { ok: true, text: parsed.text, pages: parsed.pages, truncated: parsed.truncated };
      }
      return { ok: false, reason: parsed.error ?? "抽取脚本报告失败但未给原因" };
    } catch {
      return { ok: false, reason: `抽取脚本输出不可解析（exit=${proc.exitCode}）：${(stderr || stdout).slice(0, 200)}` };
    }
  } catch (error) {
    return { ok: false, reason: `python 进程启动失败: ${error instanceof Error ? error.message : String(error)}` };
  }
}

File: backend/src/literature/pdf_text.py
# V66：PDF → 纯文本抽取（供精读卡吃全文）。
# 用法: python pdf_text.py <pdf路径> <最大字符数>
# 输出: 单行 JSON {ok, text, pages, truncated} 或 {ok: false, error}
# 纪律: 任何失败都走 ok:false + 可读 error——上游据此优雅降级回摘要，绝不静默假装有全文。
import json
import sys


def main() -> int:
    if len(sys.argv) < 3:
        print(json.dumps({"ok": False, "error": "用法: pdf_text.py <pdf> <maxChars>"}))
        return 1
    path, max_chars = sys.argv[1], int(sys.argv[2])
    try:
        from pypdf import PdfReader
    except Exception as exc:  # pypdf 未安装：报可操作的原因，不是裸 traceback
        print(json.dumps({"ok": False, "error": f"pypdf 不可用: {exc}。安装：VIRTUAL_ENV=.venv uv pip install pypdf"}))
        return 1
    try:
        reader = PdfReader(path)
        parts = []
        total = 0
        truncated = False
        for page in reader.pages:
            t = page.extract_text() or ""
            if not t:
                continue
            remain = max_chars - total
            if remain <= 0:
                truncated = True
                break
            if len(t) > remain:
                t = t[:remain]
                truncated = True
            parts.append(t)
            total += len(t)
        text = "\n".join(parts).strip()
        if not text:
            print(json.dumps({"ok": False, "error": "PDF 无可抽取文本（可能是扫描件/纯图像）"}))
            return 1
        print(json.dumps({"ok": True, "text": text, "pages": len(reader.pages), "truncated": truncated}))
        return 0
    except Exception as exc:
        print(json.dumps({"ok": False, "error": f"解析失败: {type(exc).__name__}: {exc}"}))
        return 1


if __name__ == "__main__":
    sys.exit(main())
